enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,280 @@ package hit_tracker_client
+	gstatus "google.golang.org/grpc/status"
 )
 
+const (
+	initialSliceCapacity = 4_096
+)
+
+var (
+	remoteHitTrackerTarget       = flag.String("cache.remote_hit_tracker.target", "", "The gRPC target of the remote cache-hit-tracking service.")
+	remoteHitTrackerTimeout      = flag.Duration("cache.remote_hit_tracker.rpc_timeout", 5*time.Second, "The timout to use for gRPC requests to the remote cache-hit-tracking service.")
+	remoteHitTrackerPollInterval = flag.Duration("cache.remote_hit_tracker.update_interval", 250*time.Millisecond, "The time interval to wait between sending remote cache-hit-tracking RPCs.")
+	maxPendingHitsPerGroup       = flag.Int("cache.remote_hit_tracker.max_pending_hits_per_group", 100_000, "The maximum number of pending cache-hit updates to store in memory for a given group.")
+	remoteHitTrackerWorkers      = flag.Int("cache.remote_hit_tracker.workers", 1, "The number of workers to use to send asynchronous remote cache-hit-tracking RPCs.")
+)
+
+func Register(env *real_environment.RealEnv) error {
+	conn, err := grpc_client.DialInternal(env, *remoteHitTrackerTarget)
+	if err != nil {
+		return err
+	}
+	env.SetHitTrackerFactory(newHitTrackerClient(env, conn))
+	return nil
+}
+
+func newHitTrackerClient(env *real_environment.RealEnv, conn grpc.ClientConnInterface) *HitTrackerFactory {
+	factory := HitTrackerFactory{
+		authenticator:          env.GetAuthenticator(),
+		ticker:                 env.GetClock().NewTicker(*remoteHitTrackerPollInterval).Chan(),
+		quit:                   make(chan struct{}, 1),
+		maxPendingHitsPerGroup: *maxPendingHitsPerGroup,
+		hitsByGroup:            map[groupID]*cacheHits{},
+		rpcTimeout:             *remoteHitTrackerTimeout,
+		client:                 hitpb.NewHitTrackerServiceClient(conn),
+	}
+	for i := 0; i < *remoteHitTrackerWorkers; i++ {
+		go factory.start()
+	}
+	env.GetHealthChecker().RegisterShutdownFunction(factory.shutdown)
+	return &factory
+}
+
+type groupID string
+type cacheHits struct {
+	authHeaders map[string][]string
+	hits        []*hitpb.CacheHit
+}
+
 type HitTrackerFactory struct {
-	env environment.Env
+	authenticator interfaces.Authenticator
+
+	ticker <-chan time.Time
+	quit   chan struct{}
+
+	mu                     sync.Mutex
+	maxPendingHitsPerGroup int
+	hitsByGroup            map[groupID]*cacheHits
+
+	rpcTimeout time.Duration
+	client     hitpb.HitTrackerServiceClient
+}
+
+func (h *HitTrackerFactory) NewACHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
+	return &NoOpHitTracker{}
+}
+
+func (h *HitTrackerFactory) NewCASHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
+	return &HitTrackerClient{ctx: ctx, enqueueFn: h.enqueue, client: h.client, requestMetadata: requestMetadata}
+}
+
+// For Action Cache hit-tracking, explicitly ignore everything. These cache
+// artifacts should always be served from the authoritative cache which will
+// take care of hit-tracking.
```

Comment:
nit: this comment might fit better on line 85 where NewACHitTracker() returns a NoOpHitTracker

Reply:
Done